### ISSUE TITLE
[RFC] Basic definitions

### ISSUE THREAD

**Issue by jacoscaz** [w3c/WebID#22]
Hi all. This is the first RFC thread I open and, as such, also served as a test of the RFC process itself. For more context on RFC threads, please refer to https://lists.w3.org/Archives/Public/public-webid/2023Nov/0121.html .

I kindly ask you all for your feedback on the definitions below. These are mostly @kidehen's work, although I've adapted the first one to keep closer to the current version of the spec. As per the email linked above, this is not meant to be a discussion thread. Please limit yourselves to just one comment per each of you (but feel free to have as many separate discussions about this as you want!).

1. WebID -- an identifier in the form of an HTTP URI for unambiguously naming an Agent which, when dereferenced, must always lead to a WebID-Profile Document describing the Agent named by the WebID itself.

2. WebID-Profile Document -- a document that describes an Agent named using a WebID

3. WebID-TLS -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

4. WebID-OpenIDConnect+OAuth -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

**Comment by jonassmedegaard** [comment]
I do not like that the definition of a WebID depends on the definition on WebID-Profile.

Also, the proposed wording seem ambiguous to me: Is a WebID required to never fail to dereference, in order to comply with the later "must always"?  Or conversely is the "must always" effectively a "maybe if possible"?

Seems an awful lot easier to me to define a WebID as only an identifier, without defining **at all** how that identification is validated.  And then separately define in the WebID-Profile definition that it is a mechanism to validate a WebID, by serving it as the endpoint of a WebID, and as content reiterating that same fact using RDF (without defining which serialization of RDF is used - leaving that to WebID-XXX protocol definitions).

**Comment by kidehen** [comment]
> I do not like that the definition of a WebID depends on the definition on WebID-Profile.
> 
> Also, the proposed wording seem ambiguous to me: Is a WebID required to never fail to dereference, in order to comply with the later "must always"? Or conversely is the "must always" effectively a "maybe if possible"?
> 
> Seems an awful lot easier to me to define a WebID as only an identifier, without defining **at all** how that identification is validated. And then separately define in the WebID-Profile definition that it is a mechanism to validate a WebID, by serving it as the endpoint of a WebID, and as content reiterating that same fact using RDF (without defining which serialization of RDF is used - leaving that to WebID-XXX protocol definitions).

The following WebID resolves to a profile document without introducing content-negotiation or any other distraction.

https://kingsley.idehen.net/DAV/home/kidehen/Public/YouID/link-in-bio-credentials-5/index.html#netid

The moment "#" is taken out of the naming mix, you introduce content-negotiation -- if you seek unambiguous entity naming.

**Comment by jonassmedegaard** [comment]
I don't understand the purpose of your post, @kidehen - do you mean to say that my concerns are ill-guided which is somehow supported by your referenced example and your point about use or not of hash-full URIs?

Or do you perhaps agree with my concerns and somehow your referenced example and point about use of hash emphasizes that same concern?

Or do you mean to say something completely different?

In either case, I must say that while I believe that I do understand both your example and your point about use of hash in URIs, I fail to understand what is your message here in the context of the issue discussed here.

(oh, and by my posting this second time, I probably also demonstrate that I fail to understand the declared rules of conversation for this issue...)

**Comment by kidehen** [comment]
> Also, the proposed wording seem ambiguous to me: Is a WebID required to never fail to dereference, in order to comply with the later "must always"? Or conversely is the "must always" effectively a "maybe if possible"?

I was responding to:
`Also, the proposed wording seem ambiguous to me: Is a WebID required to never fail to dereference, in order to comply with the later "must always"? Or conversely is the "must always" effectively a "maybe if possible"?`

My point:
A WebID is a name, so there has to be indirection from its sign aspect to its connotation aspect. Basically, it MUST resolve, since it is a name. 

Resolution only introduces content-negotiation (and associated headaches) if "#" isn't used as part of the signage aspect of the name -- hence my example :)

**Comment by jacoscaz** [comment]
@jonassmedegaard @kidehen as per my introduction, I kindly ask you to move any back and forth that goes beyond your respective feedback to other places (here on GitHub or on the list).

@kidehen: @jonassmedegaard has provided feedback in the form of this first comment (https://github.com/w3c/WebID/issues/22#issuecomment-1810572993). If you want to have a discussion about it - which I invite you to do! - I kindly ask you to have such discussion anywhere else but in this issue. Instead, I would appreciate your opinion on the changes I made to the first definition, that of WebID, to which I added `which, when dereferenced, must always lead to a WebID-Profile Document describing the Agent named by the WebID itself.`

Also looking forward to feedback from @acoburn, @tallted, @woutermont, @namedgraph, @csarven , @timbl just to name a few of those I've interacted with over the last few years.

**Comment by jonassmedegaard** [comment]
Thank you, @kidehen - my error was to interpret "derefeence" to mean not only getting to the referenced object (oh, it sounds so obvious now that I got it), but also fetching the contents of that object (i.e. the turtle/xml/json/whatever RDF expressions).

Now that it is no longer ambiguous to me, I have deleted that part of my comment.
If we want to preserve this conversation, you quoted my text so my original wording is still preserved within that quote.
If, however, we want to keep this thread clean to only carry comments tied directly to the original post, then we can now both delete these additional comments.

I will await @jacoscaz to hear his preferred handling of this.

**Comment by jacoscaz** [comment]
> If, however, we want to keep this thread clean to only carry comments tied directly to the original post, then we can now both delete these additional comments.

I would very much appreciate keeping this thread clean to only carry comments tied directly to the original post. Thank you!

**Comment by jacoscaz** [comment]
Don't be shy! If you have feedback, don't hesitate.

**Comment by TallTed** [comment]
> [@jacoscaz] Don't be shy! If you have feedback, don't hesitate.

I cannot speak for anyone else, but I find this "RFC thread" concept (which concept I have not encountered anywhere else in 35+ years of interaction on the 'net) difficult to work with, at best, and it does not encourage *me* to participate.

Among other things, it removes background conversation from the thread once those who are actively participating in that portion of the thread have reached some agreement, meaning that others cannot see and replay that discussion for themselves. (For immediate example, there was evidently some conversation above on this page, but it is no longer available for others to possibly be persuaded by.)

Actual discussions (NNTP or mailing list style; not like what GitHub has built, which are closer to StackOverflow-stye Q&A pages) generally preserve that thread. Allowing for such discussions only "somewhere else" means that readers of this page must seek out those other locations in order to participate. This amounts, to my mind, to an artificial roadblock to such participation.

As best as I can make out, you're actually trying to achieve something like wiki article end-results without the article history provided by a wiki (gain, I cannot recommend GitHub's reinvention of the wiki, which doesn't provide email alerts of updates nor easy comparison of arbitrary page versions), and I wonder why such a wiki workspace (or even some git-managed files!) isn't being used.

**Comment by jonassmedegaard** [comment]
I fully agree with the concerns raised by @TallTed.

**Comment by jacoscaz** [comment]
Hello @TallTed and @jonassmedegaard ,

thank you for your feedback!

> which concept I have not encountered anywhere else in 35+ years of interaction on the 'net

First of all, I should recognize that I've been alive for roughly the same time you've been an active netizen. You, along many others in this group, have seen a lot more that I did.

I've never seen RFC threads in public discourse on the internet, either, but I've used them multiple times in my professional life precisely in contexts where discussion would struggle to evolve into decision, to great success. Given the history of this group, I thought trying something different would be a worthy endeavor.

> For immediate example, there was evidently some conversation above on this page, but it is no longer available for others to possibly be persuaded by

This is intended. The point of an RFC thread is not to persuade others or be persuaded by others but to materialize a snapshot of where each of us stands on the subject of the thread. In turn, this makes locating consensus and directing further discussion where consensus is missing a lot easier. Furthermore, it gives participants a chance to make their point organically and completely, not too far from what argumentative essays allow us to do in school.

There's another tool which helps is much commonly used on the internet for a similar purpose - voting. Here on GitHub, it is often implemented by counting either +1s or reactions in the form of emojis. At this stage I am wary of calling for votes, however, as I feel that voting would disregard the technical merits of 12 years of discussion.

Is there any way I can convince you to indulge me in this attempt at using RFC threads to make some progress towards WebID 1.0? I'm not married to the concept and I'd be happy to abandon it should it prove counter-productive. Given previous experiences, however, I'd still like to try.

**Comment by acoburn** [comment]
Related to the definitions, I agree with @jonassmedegaard's point about circularity: WebID depends on WebID-Profile Document, which depends on WebID.

Perhaps the definition of WebID could be simply:

> 1. WebID -- an identifier in the form of an HTTP URI for unambiguously naming an Agent

Second, could we shorten the name of "WebID Profile Document" to just "WebID Document"? The word "Profile" has come to mean so many different things, that I don't believe it adds anything to the definition.

> 2. WebID Document -- a document that describes an Agent named using a WebID
>
> 3. WebID-TLS -- an authentication protocol for verifying credentials expressed in a WebID Document
>
> 4. WebID-OpenIDConnect+OAuth -- an authentication protocol for verifying credentials expressed in a WebID Document

Third, the WebID-TLS and WebID-OpenIDConnect+OAuth items suggest that credentials are expressed in a WebID Document. I do not believe that is strictly true. Instead, those protocols _use_ the associated WebID Document to verify credentials. That is, those documents contain data that a third party can use to verify a credential. (I don't know if we need to introduce the concept of a "credential subject" here or just leave the definition open-ended)

Finally, (minor) in the initial proposal for definitions, there is an inconsistency in the use of "a" and "an" when referring to WebID. Those could all be changed to "a".

**Comment by jacoscaz** [comment]
Excellent feedback @acoburn ,  thank you for taking the time!

**Comment by woutermont** [comment]
Apologies for the late reaction; I have been postponing a lot of things due to personal reasons.

_[Aside:] Contrary to most, apparently, I think this RFC-style gathering of feedback is a welcome addition to the endless issue threads we often see. Provided, of course, that such discussions still have a clear public platform (e.g. Github Discussions, other issues ...), and have had the time to go back and forth for some time, an RFC can be a good way for editors/chairs to have a moment-in-time assessment of where people stand towards adopting certain decisions/phrasing._

---

To aid in the understanding of my feedback, allow me to give a few higher-order semiotic definitions I tend to work with when conceptualising representation on the Web.

- **Identity**: a representation of an entity, consisting of a subset of the entity's characteristics that uniquely distinguish it from other entities. 

- **Identity Document**: is a document that represents an *Identity*.

- **Identifier**: a _minimal_ *Identity* (or identity-like representation), i.e. the _smallest_ subset of an entity's characteristics that still allows one to uniquely refer to that entity, serving as a _token_ (a sign, often a symbol) referencing the entity.

- **Dereferenceable Identifier**: an *Identifier* that can be _dereferenced_ to an *Identity Document* describing the *Entity* it represents. 

- **RDF-Dereferenceable Identifier**: a *Dereferenceable Identifier* that dereferences to an _RDF Source_.

---

Concretely then, the feedback:

> 1. WebID -- an identifier in the form of an HTTP URI for unambiguously naming an Agent which, when dereferenced, must always lead to a WebID-Profile Document describing the Agent named by the WebID itself.

This is a good definition i.m.o. It characterises the WebID as _a Dereferenceable Identifier naming an Agent and using the HTTP protocol_. 

Contrary to @jonassmedegaard and @acoburn, I think the definition of a WebID necessarily depends on that of the document it references. The document and it's structure is the dependency, the WebID the depender, even if their names might mislead us to think otherwise. Importantly, one could use the same Identity Document using multiple Identifier types (e.g. a DID that dereferences to the same document as a WebID).

> 2. WebID-Profile Document -- a document that describes an Agent named using a WebID

Agreed with @acoburn that 'Profile' could perhaps be dropped for clarity. On the other hand, I believe it is conceptually more correct to drop 'WebID' from it, since the document itself does not depend on the Identifier (but the other way around; see above). An alternative terminology avoiding both could be 'Identity Document' (which I'll use from here on).

This definition contains the core (a WebID Profile Document as _an Identity Document of an Agent_), but 
- the mention 'using a WebID' is not necessary (and creates the circular dependency @jonassmedegaard wants to avoid); and
- this definition needs to be extended to include the expected structure of the document: an RDF document with as `foaf:primaryTopic` the Agent, which contains information that can _authenticate_ the Agent (e.g. Cert. Key or OIDC Issuer). Without this, the WebID spec amounts to little, since the power of it lies precisely in what we can expect to find when dereferencing a WebID. This can depend on the mechanism, but needs to be mentioned _in abstracto_ in the Identity Document definition (e.g. as a [specification-profile](https://www.w3.org/TR/spec-variability/#subdivision-profile)-based extension point of which WebID-TLS and WebID-OIDC are extension profiles).

> 3. WebID-TLS -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

This seems wrong. As @acoburn pointed out, the use case is not "verifying something that contained in the Identity Document", but rather "verifying the Agent's Identity (i.e. authenticate the Agent) _using_ something _from_ the Identity Document".

As stated in re 2., I would explicitly define WebID-TLS as an extension profile of WebID.

_**[Suggestion] WebID-TLS**: an extension profile of WebID, defining an proof-of-posession authentication protocol that authenticates Agents by matching a given private key to the public key declared in their Identity Document._

> 4. WebID-OpenIDConnect+OAuth -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

Likewise, as in 3, both re definition as re extension profile. 

Minor: I would shorten to WebID-OIDC, and I don't get why OAuth should be in the name here.

_**[Suggestion] WebID-OIDC**: an extension profile of WebID, defining an OIDC-based authentication protocol that authenticates Agents using the OpenID Provider declared in their Identity Document._

Hope to have been of help 🙂

**Comment by jacoscaz** [comment]
@woutermont another example of excellent feedback. Thank you!

> Provided, of course, that such discussions still have a clear public platform (e.g. Github Discussions, other issues ...), and have had the time to go back and forth for some time, an RFC can be a good way for editors/chairs to have a moment-in-time assessment of where people stand towards adopting certain decisions/phrasing

This is exactly my intention. So far, the two RFC thread focus on matters that have been touched upon many times. Please do let me know the moment I call for an RFC on something that would have deserved more discussion in preparation for that.

**Comment by TallTed** [comment]
I think it's important to start linking these RFC threads to (some of, at least one of) their precursor discussion threads, to both (a) insure that such discussion threads *do* exist and (b) provide people who may be new to the topic of the RFC thread with a way to catch up on "the discussion to date".

**Comment by melvincarvalho** [comment]
> Hi all. This is the first RFC thread I open and, as such, also served as a test of the RFC process itself. For more context on RFC threads, please refer to https://lists.w3.org/Archives/Public/public-webid/2023Nov/0121.html .
> 
> I kindly ask you all for your feedback on the definitions below. These are mostly @kidehen's work, although I've adapted the first one to keep closer to the current version of the spec. As per the email linked above, this is not meant to be a discussion thread. Please limit yourselves to just one comment per each of you (but feel free to have as many separate discussions about this as you want!).
> 
>     1. WebID -- an identifier in the form of an HTTP URI for unambiguously naming an Agent which, when dereferenced, must always lead to a WebID-Profile Document describing the Agent named by the WebID itself.

I've always said

A WebID is an HTTP URI that denotes an Agent

(any similar wording is fine)

> 
>     2. WebID-Profile Document -- a document that describes an Agent named using a WebID

I dont think this is right.  Any number of documents could describe a WebID

The profile document is obtained from the WebID (e.g. after removing the #)

> 
>     3. WebID-TLS -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

Could the key be in another document?

> 
>     4. WebID-OpenIDConnect+OAuth -- an authentication protocol for verifying credentials expressed in an WebID-Profile Document

Slight nit: having "-" and "+" in the same thing doesnt scan well.

All looks fine, in general

**Comment by melvincarvalho** [comment]
> There still appears to be confusion about what a WebID is, so here's another collection of anecdotes that hopefully brings clarity to the issue at hand:
> 
> A passport number and a passport document used by a traveler.
> 
> A passport number is an identifier that unambiguously denotes a person -- this is what a WebID is all about. A passport is a document comprising credentials pegged to a subject denoted by a passport number -- this is what a WebID Profile document is all about.
> 
> Passport authentication is an act of verifying credentials associated with a passport number.
> 
> You can also think about a driver's license number and a driver's license document.
> 
> A WebID, Profile Document, and Authentication Protocols are 3 distinct things that are loosely related.

Very quickly on this.  I dont think it's right.  A passport number can be written on a piece of paper, which doesnt mean that paper is a passport.  A WebID can be used to obtain a document, but a passport number cant.  The WebID and the WebID document are not loosely coupled.

**Comment by kidehen** [comment]
> > ```
> > 2. WebID-Profile Document -- a document that describes an Agent named using a WebID
> > ```
> 
> I dont think this is right. Any number of documents could describe a WebID
> 

Alternatively:

WebID-Profile Document -- a document about an Agent named using a WebID.

**Comment by woutermont** [comment]
I must express how flabbergasted I am be the sheer incapability of people to follow a simple request of the issue starter (and chair of this CG). @jacoscaz has repeatedly asked to treat this issue as an RFC thread, to limit ourselves to one comment each, and to have discussions elsewhere. Yet the majority of responders somehow feel that this is not addressed to them. 

Can we please remove the responses (including this one) and continue in line with what was asked?

**Comment by melvincarvalho** [comment]
Deleting items can be concerning for some, as it potentially limits the historical record. This is why a number of people favor working with mailing lists, where deletion is not an option, over platforms like GitHub.

**Comment by jacoscaz** [comment]
@woutermont I very much appreciate your comment and support. Thank you!

I should also add, though, that RFCs are just a tool that I am experimenting with as previous experiences have brought me to believe it _may_ help with some of the group's struggles in finalizing the spec. I am absolutely open to feedback and I would have no issue dropping RFC threads entirely should the majority of active participants find them unpalatable. So far there seems to be a 50%/50% split.

For transparency: I've renamed this issue to "Basic Definitions", dropping the "[RFC]" prefix. Nonetheless, I still kindly ask participants to please approach this conversation in the spirit of an RFC thread, at least insofar as willing or capable.

**Comment by woutermont** [comment]
To clarify: by 'remove' I mean to hide them, which @jacoscaz should be able to do. I fully agree that permanent deletion is not the way to go.

**Comment by jacoscaz** [comment]
Hi all, 

Based on all the feedback above, I've pieced together the following definitions:

- **WebID**: an HTTP URI that denotes an agent and dereferences to an Identity Document describing such agent.

- **Identity Document**: an RDF document that describes an Agent, consisting of at least one statement identifying the agent as the document's `foaf:primaryTopic`.

Of course, as with most compromises, these are unlikely to make everyone as happy as everyone could be. The goal, however, is to have something that makes nobody, or as few of us as possible, unhappy to the point of considering them a no-go.

What do you think? @woutermont @jonassmedegaard @melvincarvalho @kidehen @acoburn @TallTed 

_Please be explicit in terms of what you consider a no-go vs. what you'd rather do differently but are ultimately ok with._

**Comment by melvincarvalho** [comment]
> WebID: an HTTP URI that denotes an agent and dereferences to an Identity Document describing such agent.

+1 seems reasonable at first glance

> Identity Document: an RDF document that describes an Agent, consisting of at least one statement identifying the agent as the document's foaf:primaryTopic.

-0 I understand it's a compromise solution, and that it's imposslbe to please everyone, though it's tied to FOAF which I think is no longer maintained.

I would not implement this personally, though it's a reasonable way to go.

I would rather use one of [inverse terms](https://github.com/w3c/WebID/issues/24) either ``foaf:isPrimaryTopicOf` or `schema:mainEntityOfPage`.  I would NOT use `foaf:primaryTopic` myself, I dont think.  But may interop with WebIDs that do.

**Comment by melvincarvalho** [comment]
A more general formulation of (2)

> Identity Document: an HTTP document that describes an Agent, consisting of at least one statement tying the agent to the Identity Document

Then: explain the terms Agent and Primary Topic

Note that in FOAF a group can be an Agent too, and groups can have sub members.  Would those sub members be WebID's too?  That is for the CG to decide.

**Comment by jacoscaz** [comment]
> which I think is no longer maintained.

@danbri @libbymiller what is the state of FOAF? The http://www.foaf-project.org website doesn't seem to work anymore and http://xmlns.com/foaf/0.1/ has a few warnings mentioning the fact that it's temporarily serving an older version as they work on restoring from backups. https://www.iptc.org/thirdparty/foaf/ has what I _think_ is the most up-to-date version of the spec.

**Comment by libbymiller** [comment]
hi, I'm investigating!

**Comment by jacoscaz** [comment]
thank you @libbymiller for your time, much appreciated!

**Comment by TallTed** [comment]
Also note https://web.archive.org/web/20211024073639/http://xmlns.com/foaf/spec/

**Comment by jacoscaz** [comment]
Thank you @danbri for following up - much appreciated!

**Comment by woutermont** [comment]
@jacoscaz, to get back to the draft definitions you provided: they are both workable drafts i.m.o. 

Two minor points (preferable for me):

- I would use "about" rather than "describes". Since the WebID spec is a tool to enable a lot of other mechanisms (of which description can be one), I'd rather leave it open if the Identity Document actually provides any kind of "description" except in its broaderst but trivial sense (it contains at least a triple with the Agent as a node).

- The `foaf:primaryTopic` link is a necessary but foremost practical part of the Identity Document. It does not really have to be in the main definition.

One major point (breaking for me): the core of the Identity Document is to provide a hook for authentication mechanisms. This should really be part of the definition.

Applied to your draft, this gives the following proposal.

> * **WebID**: an HTTP URI that denotes an Agent and dereferences to an Identity Document <del>describing such</del> <ins>about that</ins> Agent.
> * **Identity Document**: an RDF document <del>that describes</del> <ins>about</ins> an Agent, <ins>providing third parties with one or more ways to authenticate that Agent.</ins>
> <ins>An Identity Document MUST</ins> consist<del>ing</del> of at least one statement identifying the Agent as the Document's `foaf:primaryTopic`, <ins>and one or more statements indicating an authentication mechanism as described by a WebID extension profile</ins>.

**Comment by jacoscaz** [comment]
Thank you @woutermont for your feedback!

I _personally_ like where you are going. However, given the note on the extensibility of specs, doesn't the above imply that a WebID implementation MUST actually implement at least one Extension Profile? If so, this feels like something that would better be stated explicitly.

On the flip side, not everyone in the group might agree on binding the primary spec to the implementation of one of its Extension Profiles. @melvincarvalho @kidehen @jonassmedegaard @acoburn ?

**Comment by woutermont** [comment]
> Given the note on the extensibility of specs, doesn't the above imply that a WebID implementation MUST actually implement at least one Extension Profile? If so, this feels like something that would better be stated explicitly.

Yes, of course. A WebID is useless as long as the agent can't prove that it is actually theirs. Otherwise I could just use your WebID to impersonate you.

The last sentence of my proposal already contains the explicit mention of that requirement. ('_MUST consist of ... and one or more statements indicating an authentication mechanism as described by a WebID extension profile_')

**Comment by jacoscaz** [comment]
@woutermont apologies, I am reading through a lot of activity and got two different comments mixed up in my head. Sorry! Please disregard my comment.

Nonetheless, the question remains of whether the group agrees on binding implementations of the spec to supporting at least one extension profiles.

@jonassmedegaard consider opening a dedicated thread for that specific question. For me, a WebID MUST dereference to an identity document via HTTP(S). So yes, WebID _requires_ HTTP. I am very afraid of what might happen if that also becomes a point of contention.

**Comment by woutermont** [comment]
I think that's precisely what @jonassmedegaard is challenging (the HTTP question was just an analogy).

But I believe the analogy is wrong. 

- URI does not require HTTP because its purpose is not to transfer hypertext messages. HTTP does not require HTML because its purpose is not to provide a way to write linked markup documents. The correct analogy here would be that HTTP does also not require WebID, because its purpose is not to authenticate agents.

- On the other hand, HTTP _does_ require URI, since to perform its purpose of transferring hypertext it needs a way to address hypertext resources. HTML kind of requires HTTP (or at least one transfer protocol supported by the browser), because to fulfill its purpose of writing linked markup documents, it needs at least one way for browsers to follow those links. Similarly, WebID needs at least one authentication mechanism in order to fulfil its purpose.

**Comment by jacoscaz** [comment]
> Yes, of course. A WebID is useless as long as the agent can't prove that it is actually theirs. Otherwise I could just use your WebID to impersonate you.

> Similarly, WebID needs at least one authentication mechanism in order to fulfill its purpose.

IMHO, a version of one of these should end up in the introductory section of the spec.

**Comment by melvincarvalho** [comment]
> Thank you @woutermont for your feedback!
> 
> I _personally_ like where you are going. However, given the note on the extensibility of specs, doesn't the above imply that a WebID implementation MUST actually implement at least one Extension Profile? If so, this feels like something that would better be stated explicitly.
> 
> On the flip side, not everyone in the group might agree on binding the primary spec to the implementation of one of its Extension Profiles. @melvincarvalho @kidehen @jonassmedegaard @acoburn ?

FWIW I'm good with any of the definitions at this point.  I agree with your sentiment of "not letting the perfect be the enemy of the good".  And also there has been enough progress to unblock all my work.

I find it a useful though experiment to consider an RDF document where a user makes a comment.  That user will have a webid, and the webid will be in that document.  However the document is not the profile of the webid.  If the definitions all work in that scenario, they are probably ticking a lot of boxes.

**Comment by csarven** [comment]
Responding only to the definitions themselves:

The current WebID 1.0 ED already defines WebID and WebID Profile Document under the Terminology section. They can be clarified with the understanding that the intention is preserved. As for the term "unambiguously", I think that works both in layperson terms and specifically, but I also think that it need not be said because it is a given as per Web Architecture. The URI owner associates a URI to a resource, and there is a social expectation (SHOULD) that they preserve what it identifies.

Regarding the definitions of WebID-TLS, WebID-OpenIDConnect+OAuth:

They don't belong in the WebID 1.0, unless it is relevant for Conformance or Considerations. WebID-OpenIDConnect+OAuth in particular is only speculated - there is no specification or requirement, or even a work item for it.

---

WebID is useful and fulfils the purpose of identifying an agent ("Activity X has actor Y") independently from being employed for authentication.

---

I believe the current intention of WebID + WebID Profile Document is analogous (if not identical) to "RDF-Dereferenceable Identifier". It is technically possible to have a non-RDF-document that describes a given WebID and still be consistent with the Web Architecture. But that's out of scope for what's intended to be detailed in the WebID specification as I see it.

**Comment by woutermont** [comment]
> [@csarven:] WebID is useful and fulfils the purpose of identifying an agent ("Activity X has actor Y") independently from being employed for authentication.

Could you elaborate? 

If there is no authentication mechanism, a WebID and its WebID Document are nothing more than just another URI and the RDF Document it derefences to. We don't need a new spec to use those; we already have RDF: If all one needed was a URI pointing to some RDF Document, one could simply ask the user for such a URI.

**What we need, and what WebID (should) provide(s), is a way to ensure that URI and the information in that RDF Document are indeed under the control of the Agent we are interacting with.**

**Comment by kidehen** [comment]
> * The `foaf:primaryTopic` link is a necessary but foremost practical part of the Identity Document. It does not really have to be in the main definition.

Yes!

**Comment by kidehen** [comment]
> Applied to your draft, this gives the following proposal.
> 
> > * **WebID**: an HTTP URI that denotes an Agent and dereferences to an Identity Document ~describing such~ about that Agent.

How about "resolves to" rather than "dereference" for broader readability?

> > * **Identity Document**: an RDF document ~that describes~ about an Agent, providing third parties with one or more ways to authenticate that Agent.
> >   An Identity Document MUST consist~ing~ of at least one statement identifying the Agent as the Document's `foaf:primaryTopic`, and one or more statements indicating an authentication mechanism as described by a WebID extension profile.

That works to, since a profile document is about an agent named by a WebID.

**Comment by kidehen** [comment]
> What we need, and what WebID (should) provide(s), is a way to ensure that URI and the information in that RDF Document are indeed under the control of the Agent we are interacting with.

The problem with this riddle is that its solution depends on where you start. 
If you start from the point of trying to verify what a WebID is, via machine-computable content in a document,  a broadly accepted answer isn't unattainable.

if you start from the notion of a WebID-Profile document, where the prime purpose is identification, a broadly accepted answer is attainable -- since this document will comprise an entity relationship graph where entity relationship type semantics are machine-computable (or readable).

My rolling analogy:
Trying describing the purpose of a Passport Number vs the Passport itself. What's makes the Passport Number an Identifier that denotes a Person unambiguously?

**Comment by woutermont** [comment]
> [@kidehen:] How about "resolves to" rather than "dereference" for broader readability

I know they are often used interchangeably in common language, but in specifications "resolving" often specifically means disambiguating the identifier (e.g. a relative URI reference is resolved to an absolute one), while "dereferencing" always clearly incorporates the extra step of actually trying to retrieve the meaning of that disambiguated (resolved) identifier (e.g. fetching a representation of the resource identified by the URI). That's why I prefer "dereference", but I won't press much if the group decides otherwise.

**Comment by woutermont** [comment]
Re purposes, I'm going to appropriate your analogy, @kidehen. I think there is some value in the distinction you make, but just as how a Passport Number only has purpose relative to the Passport, a WebID only has purpose relative to the WebID Document. The purpose of both identifiers comprises mostly of refering to their respective document (in a dereferenceable way, to be practically useable). They stand, as it were, in service of another (higher) purpose. Thus, when I say "the purpose of a WebID," I actually mean "the purpose of a WebID Document, and therefore the indirect purpose of a WebID" (but that is quite long to write/read).

**Comment by csarven** [comment]
>>WebID is useful and fulfils the purpose of identifying an agent ("Activity X has actor Y") independently from being employed for authentication.

>Could you elaborate?

Wouter, I mean as far as what the definition of WebID (the identifier) includes/entails, i.e., an HTTP URI based identifier to denote agents. Referencing a WebID (in an RDF statement for example) would be a separate use case to an authentication mechanism using WebID.

```
<https://linkedresearch.org/annotation/csarven.ca/%23i/87bc9a28-9f94-4b1b-a4b9-503899795f6e>
  as:actor <https://csarven.ca/#i> .

<https://csarven.ca/#i>
  foaf:knows <https://wouter.termont.online/> .
```

**Comment by woutermont** [comment]
@csarven, I get what you are saying, but my reply remains the same: you don't need WebIDs for that. All the semantics you need is in the RDF there: some node has an actor identified by `https://csarven.ca/#i`, which knows an node identified by `https://wouter.termont.online/`, and the two last nodes are Agents (by courtesy of `foaf:knows` semantics).

As far as I am concerned, we are not writing a spec that, at its core, does nothing more than what we already can do without it 🤔  What a WebID can and should provide _extra_, is that on interpretation of those statements, you know you can, if so desired, verify that the agent identified by `https://csarven.ca/#i` is really you, the person Sarven Capadisli, and the one identified by `https://wouter.termont.online/` is really me, the person Wouter Termont.

**Comment by kidehen** [comment]
> They stand, as it were, in service of another (higher) purpose. Thus, when I say "the purpose of a WebID," I actually mean "the purpose of a WebID Document, and therefore the indirect purpose of a WebID" (but that is quite long to write/read).

Yes. 

But when it comes to specs, most of the work lies in spelling out what's ends up being long to write and lengthy to read. 

You've actually summed up, quite nicely, the real problem with the current spec i.e., its structural organization and content is oriented towards a particular authentication protocol -- for which TLS is the only implementation. 

The protocol:
Look-up a WebID in profile documents by navigating an RDF-based entity relationship graph that's been crafted from  specific entity relationship types. 

Back then, FOAF was the natural source of entity type (classes) and relationship types (properties) for the aforementioned protocol. 

Unbeknownst to many, we went through a similar process (long before Paris) re evolution of the WebID-Profile doc from RDF-XML to RDF-Turtle. That too, was lengthy and arduous -- despite all the well established mass adoption problems that plagued RDF-XML.

**Comment by melvincarvalho** [comment]
> Edit: I am curious, though, as to what other people think of this. I also realise that I do not want to make this a breaking point, because even if not mandatory, every WebID Document that is actively used will end up containing an authentication mechanism anyhow.

A little historical perspective.  This was the mainstream view at the start and for the first 6 years of webid which was called foaf+ssl.

The idea, which was and still is the mainstream prevailing view, that identity and authentication should be closely coupled to provide a useful service.

In 2014 I have a fringe minority idea, which I presented at the RWW+WebID workshop, sitting between Timbl and Henry.

I made a point, which was quite heretical, that identity without authentication is still valuable.  It was one of those things that I expected to get shot down immediately because it was radical and a big departure from what we had done for many years.

My argument was that 3 widely adopted identity systems operate without verification.  Namely, email, postal addresses and mobile phones.  You can see who is identified, but you might not verify that in every case.  But that verification can be added on to each system.

Again, I was surprised that this heretical view was not shot down before I got to the end.  Henry was shaking his head.  And to my absolute astonishment Timbl listened to the idea, even liked it, and so did most of the others in the room.  Henry argued that if you separate FOAF+SSL the identity page isnt a spec, because it would just be 2 pages.  But timbl said that is a good thing, that we define identity as separate from authentication.  In the end Henry conceded and said he'd split the specs, so that it was neither tied to FOAF nor to SSL (ie TLS).

I'm still to this day amazed that things worked out this way, because everyone else was tying together identity and auth.  So that's something that makes WebID different.

It was then decided that we need one MUST serialization so that everything would work together.  And Turtle was chosen.  I didnt like that because I hadnt implemented turtle, but I got my webid working with .ttl the same day.

Why turtle?  Well in 2014 we thought it was a good idea.  It was a good group (agile, low bureaucracy, high development), it was a "pure" and clean data structure, it was standarized and we had a chance to show people what it could do via interop if enough people decided to take it up.  In short, it was a bet.  10 years later it's not a bet that paid off, because developers have gravitated to JSON-LD.

So now, 10 years later, we have to acknowledge what we've learnt.  And there is enough water under the bridge, to modernize the spec.  The simple approach being to keep webid as a common definition.  Then Turtle=MUST as a profile that implements that common WebID definition.  Similarly, we can have a JSON-LD profile for those that want it.

Authentication today can be done in a number of methods and the original WebID-TLS barely works in browsers, but there are modern alternatives.

So while I get  your point that identity without authn is "useless", indeed, I think that is the prevailing mainstream view.  Hope I have given an explanation of another approach, some history of the path to get there.  And possibly a future path.  I dont want to discourage your convictions.  But just suggest that if there is a loosely coupled spec that contains identity, serialization and auth, then it can live in one document, or common parts can be lifted into other documents and provide points of flexibility, which may end up giving everyone what they want.  IMHO we are 90% of the way there, whether the final 10% happens inside the CG or outside we can all still work, but inside would be preferable.

**Comment by kidehen** [comment]
> that identity without authentication is still valuable. It was one of those things that I expected to get shot down immediately because it was radical and a big departure from what we had done for many years.

That's fundamental to the use of an HTTP URI as a Name.  That's the fundamental basis of Linked Data Principles too. 

Like WebID, Linked Data Principles also ended up being challenged by the use of HTTP URIs as Names. 

Denotation --> Connotation indirection packed into a Name isn't always immediately obvious. 

A "#" based indexical tacked on to a URL pulls off the trick, but it simply isn't generally understood. 

For additional historic perspective:

A major reason for JSON-LD being subjugated in the spec arises from the fact that "#" based relative URIs weren't supported until JSON-LD 1.1 . That's also why, under the right circumstances, it should be easier to get it on level pegging with Turtle regarding the WebID spec.

**Comment by kidehen** [comment]
> My argument was that 3 widely adopted identity systems operate without verification. Namely, email, postal addresses and mobile phones. You can see who is identified, but you might not verify that in every case. But that verification can be added on to each system.
> 
> Again, I was surprised that this heretical view was not shot down before I got to the end. Henry was shaking his head. And to my absolute astonishment Timbl listened to the idea, even liked it, and so did most of the others in the room. Henry argued that if you separate FOAF+SSL the identity page isnt a spec, because it would just be 2 pages. But timbl said that is a good thing, that we define identity as separate from authentication. In the end Henry conceded and said he'd split the specs, so that it was neither tied to FOAF nor to SSL (ie TLS).

Yes, and of course TimBL would support that, since that's a reiteration of what he already knows.

**Comment by woutermont** [comment]
Thanks a lot for the historical annecdote, @melvincarvalho!

However, I don't want ot be mischaracterised as saying that identity without authentication is useless. It is absolutely not, and I think we all agree on that.

My point, on the other hand, was that the identity-without-authentication part is already handled adequately by URIs. We don't need a new spec for that. If I want to state something in RDF about you, I can do that with a URI that is not a WebID as well.

Thus, for the WebID spec to have purpose, it should add something to the existing ecosystem, namely the ability to check that, when I use that URI, I am indeed talking about you, and not someone (or something) else.

**Comment by kidehen** [comment]
> However, I don't want ot be mischaracterised as saying that identity without authentication is useless. It is absolutely not, and I think we all agree on that.
Thus, a  WebID should be able to exist in a spec without being conflated with a Profile and associated protocol. 

We only have a problem right now because the spec was a retrospective deconstruction that started with what was then called FOAF+SSL -- an authentication protocol. 

It had an aborted start as FOAF based Profile as the basis for "Getting Yourself a URI" , and the FOAF+SSL was tacked on en route to the WebID spec we have today. 

Anyway, I think we are getting closer to the following:

A better spec there separation of concerns are clearer for more parties, rather than specific parties. 

What does that look like? 

1. PRs to fix what's wrong with the current spec -- my preference
2. Fork of the current spec into a separate spec comprising the required fixes -- primarily putting Turtle and JSON-LD on a level playing field re WebID-Profie document types and minimum expectations on the parts of publishers

**Comment by woutermont** [comment]
We are straying far fro the point of discussion, I'm afraid. The last real question was: do we or do we not think that a WebID Document should contain at least one triple indicating an authentication mechanism/profile?

Am I correct in deriving from the previous comments that @melvincarvalho and @kidehen don't see this as a necessity?

What do other people think?

**Comment by kidehen** [comment]
> We are straying far fro the point of discussion, I'm afraid. The last real question was: do we or do we not think that a WebID Document should contain at least one triple indicating an authentication mechanism/profile?

A WebID-Profile document should, but I would characterize the need as follows:

One of more statements for establishing what type of entities the document is about.  

I refer to entities, rather than a single entity, since a WebID-Profile document could be about many entities that are each unambiguously named using an HTTP URI comprising an "#" based indexical.

**Comment by woutermont** [comment]
> Nothing to do with an authentication protocol, which is a different matter.

But that is precisely the question at hand...

**Comment by jacoscaz** [comment]
_/chair hat off_

> The last real question was: do we or do we not think that a WebID Document should contain at least one triple indicating an authentication mechanism/profile?

> One of more statements for establishing what type of entities the document is about.

These two read to me as being about entirely different things. The first is about authentication mechanisms (for example, the triples for WebID-TLS), the second appears to be about entity types (for example, `<webid> a foaf:Agent`). @kidehen  Is my reading wrong?

@woutermont My personal answer is yes - each WebID implementation should implement at least one Extension Profile and, therefore, the respective documents should contain at least one authentication-related triple (or however many are needed for the Extension Profile that requires the least amount of triples).

**Comment by kidehen** [comment]
If so, I am not in agreement. 

A profile document can be _solely_ about one or more agents whose entity type is both human and machine-computable -- by way of content processing (a/k/a reading).

**Comment by kidehen** [comment]
> These two read to me as being about entirely different things. The first is about authentication mechanisms (for example, the triples for WebID-TLS), the second appears to be about entity types (for example, `<webid> a foaf:Agent`). @kidehen Is my reading wrong?

No, your interpretation is correct. 

**Key point of clarification:**
Triples in a profile document that determine entity type also have the dual effect of providing important information to an authentication protocol. Likewise, they also inform other HTTP agents that lookup a WebID for other purposes. 

I have applications that demonstrate these other aspects of a WebID that cover things like:

Automagically, associating HTTP client application sessions with specific WebIDs en route to properly partitioned content persistence for reuse and recall purposes e.g., figuring out a folder for content storage associated ChatGPT sessions.

**Comment by jacoscaz** [comment]
_/chair hat off_

I should also add that, regardless of the fact that I see WebID as fundamentally being about authentication, I do see the value in (I think) @kidehen 's and @webr's proposition and I would not consider it a blocker if the group decided to completely extricate authentication from the primary spec, leaving it entirely to Extension Profiles. Practically speaking, it wouldn't make a difference in all the use cases for which I'm considering using WebID.

**Comment by kidehen** [comment]
> I should also add that, regardless of the fact that I see WebID as fundamentally being about authentication, 

More specifically: 
It is a spec derived from an authentication protocol called FOAF+SSL. 

> I do see the value in (I think) @kidehen 's and @webr's proposition and I would not consider it a blocker if the group decided to completely extricate authentication from the primary spec, leaving it entirely to Extension Profiles.

Yes, that will get us out of this problem. The we would end up with Profile Document specifications for:

1. JSON-LD
2. RDF-Turtle

Preferably, this would be the last resort, if we can't get there by applying PRs the current spec.

**Comment by woutermont** [comment]
> Practically speaking, it wouldn't make a difference in all the use cases for which I'm considering using WebID.

I would be very interested in use cases of WebID that do not require authentication to be possible.

**Comment by kidehen** [comment]
> I do see the value in (I think) @kidehen 's and @webr's proposition and I would not consider it a blocker if the group decided to completely extricate authentication from the primary spec, leaving it entirely to Extension Profiles.

Yes, that will get us out of this problem; and then we'll end up with Profile Document specifications for:

1. JSON-LD
2. RDF-Turtle

Preferably, this would be the last resort, if we can't get there by applying PRs the current spec.

JSON-LD today is playing the role of Turtle (on the back of the LOD Cloud) regarding pragmatic tweaks to the WebID spec. 

JSON-LD is everywhere, and Turtle is very much a powerful but niche notation.  I have the following world-view I tend to share about both:

Turtle is for thinking while JSON-LD is for printing -- with regards to entity relationship graphs. 

**JSON-LD Issue Backgrounder:**

JSON-LD can now be used for crafting RDF on par with Turtle. Support for hash-based relative HTTP URIs was the real technical problem that initially lead to its resistance in this spec.  For instance, it would be inconceivable (and rightly so) for TimBL to support an RDF notation that _explicitly failed_ to support hash-based HTTP URIs. 

Luckily, JSON-LD 1.1 fixed that showstopper.

**Comment by kidehen** [comment]
> I would be very interested in use cases of WebID that do not require authentication to be possible.

Looking a WebID to determine pathways to content published by its referent.

**Comment by woutermont** [comment]
> > I would be very interested in use cases of WebID that do not require authentication to be possible.
> 
> Looking a WebID to determine pathways to content published by its referent. 
> 

Which you will never know are actually published by its referent unless authentication is possible.

**Comment by woutermont** [comment]
> sorry, the articles could be signed. can happen at any level

Signed by who? Verifiable by whose key?

As long as you are not sure of a single piece of information that it comes from the real Agent, you are nowhere.

**Comment by jacoscaz** [comment]
> I would be very interested in use cases of WebID that do not require authentication to be possible.

@woutermont for me it's the opposite: all of my current use cases require authentication, meaning that I would always implement at least one Extension Profile related to authentication. Whether the primary spec depends on the implementation of one such profile or not doesn't make a difference in my case - I would do so anyway. 

That said, if completely extricating authentication from the primary spec and mandating `<webid> a foaf:Agent` on top of `<> foaf:primaryTopic <webid>` is all it takes to reconcile the group, I would personally be all for it.

Furthermore, I can think of use cases in which trust in the document doesn't equate to trust in the Agent as described by the document and vice-versa. For example, in the case of WebID-TLS, the certificate used for serving the document over HTTPS is different from the certificate used by the Agent as a form of authentication and described in terms of exponent and modulus in the document itself. In this case, trust in the document might originate from something else (for example, a private CA on a corporate server). This means that the Agent doesn't always have to demonstrate control over the document.

Does this make sense? It's getting late, time to disengage for the day :)

**Comment by kidehen** [comment]
> As long as you are not sure of a single piece of information that it comes from the real Agent, you are nowhere.

IMHO, that's too coarse-grained a perspective. 

Identity authenticity concern arises when privileged operations need to be performed. Every interaction with a WebID isn't about performing a privileged operation.

**Comment by woutermont** [comment]
> Trust in the document might originate from something else (for example, a private CA on a corporate server). This means that the Agent doesn't always have to demonstrate control over the document.

Control over the document is implied by the presence of the statement linking the WebID to the TLS certificate.

**Comment by woutermont** [comment]
> Identity authenticity concern arises when privileged operations need to be performed. Every interaction with a WebID isn't about performing a privileged operation.

Authentication is not only necessary to perform privileged operations. It is also necessary to ascertain authority over certain information. **Once more: information without authority can be expressed in any RDF document. If authority is not needed, a WebID Document is not needed.**

But I feel like I am repeating myself, so I will rest my case for now.

**Comment by kidehen** [comment]
> Authentication is not only necessary to perform privileged operations. It is also necessary to ascertain authority over certain information.

True, but all interactions with a WebID don't fall into that scenario.

What's wrong with [this WebID](https://kingsley.idehen.net/DAV/home/kidehen/Public/YouID/link-in-bio-credentials-5/index.html#netid), based on how I've used it right here?

**Comment by woutermont** [comment]
> What's wrong with [this WebID](https://kingsley.idehen.net/DAV/home/kidehen/Public/YouID/link-in-bio-credentials-5/index.html#netid), based on how I've used it right here?

Well, for starters, it serves HTML, and I can't request it with `Accept: text/turtle`, so no spec-compliant WebID application will know what to do with it (no, alternate links and data islands are not in the spec) 🤷‍♂️ Moreover, since it is served as `text/hml`, the your fragment-based WebID actually points to a part of the document or some state information for in-page 
scripts, not a real-life person (yes, this is an often made mistake: the semantics of the fragment identifier are determined by the media type definition, so the "trick with the hash" is often played wrong).

All that is nit-picking about syntax though. Semantically, I don't think there's a problem with it. There's a public key in there, so that's at least one way in which I could ascertain that you actually have control over the other statements in there, if I really needed to.

I will rest my case though, because I feel like I'm the only one striving for the MUST re authentication. I grant that, if trust in the data is not required, a WebID document can still provide information about an entity (just like any other RDF document). If you all want to call any URI people give to eachother to refer to themselves a WebID, I won't lie awake at night.

**Comment by kidehen** [comment]
> Well, for starters, it serves HTML, and I can't request it with `Accept: text/turtle`, so no spec-compliant WebID application will know what to do with it (no, alternate links and data islands are not in the spec) 🤷‍♂️ Moreover, since it is served as `text/hml`, the your fragment-based WebID actually points to a part of the document or some state information for in-page
> scripts, not a real-life person (yes, this is an often made mistake: the semantics of the fragment identifier are determined by the media type definition, so the "trick with the hash" is often played wrong).

So it would be fine if I enabled a re-write rule to handle your request for Turtle?

**Comment by woutermont** [comment]
> So it would be fine if I enabled a re-write rule to handle your request for Turtle?

If by 'fine' you mean 'spec-compliant', then sure.

**Comment by melvincarvalho** [comment]
> I cannot speak for anyone else, but I find this "RFC thread" concept (which concept I have not encountered anywhere else in 35+ years of interaction on the 'net) difficult to work with, at best, and it does not encourage _me_ to participate.

Agreed. Despite giving RFCs a fair evaluation, my experience aligns with the sentiment that they do not effectively facilitate consensus-building in our discussions.

**Comment by jacoscaz** [comment]
Closed by #60.
